runtime/src/accounts_index.rs

```diff
@@ -173,14 +173,154 @@ impl<T: 'static + Clone + IsCached> WriteAccountMapEntry<T> {
     }
 }
 
+use bv::BitVec;
+
 #[derive(Debug, Default)]
+pub struct RollingBitField {
+    max_width: u64,
+    min: u64,
+    max: u64, // exclusive
+    bits: BitVec,
+    count: usize,
+}
+// functionally similar to a hashset
+// Relies on there being a sliding window of key values. The key values continue to increase.
+// Old key values are removed from the lesser values and do not accumulate.
+impl RollingBitField {
+    pub fn new(max_width: u64) -> Self {
+        assert!(max_width > 0);
+        assert!(max_width.is_power_of_two()); // power of 2 to make dividing a shift
+        let bits = BitVec::new_fill(false, max_width);
+        Self {
+            max_width,
+            bits,
+            count: 0,
+            min: 0,
+            max: 0,
+        }
+    }
+
+    fn get_address(&self, key: &u64) -> u64 {
+        Self::calc_address(self.max_width, key)
+    }
+
+    // find the array index
+    fn calc_address(max_width: u64, key: &u64) -> u64 {
+        key % max_width
+    }
+
+    fn check_range(&self, key: u64) {
+        assert!(
+            self.count == 0
+                || (self.max.saturating_sub(key) <= self.max_width as u64
+                    && key.saturating_sub(self.min) < self.max_width as u64),
+            "out of range"
+        );
+    }
+
+    pub fn insert(&mut self, key: u64) {
+        self.check_range(key);
+        let address = self.get_address(&key);
+        let value = self.bits.get(address);
+        if !value {
+            self.bits.set(address, true);
+            if self.count == 0 {
+                self.min = key;
+                self.max = key + 1;
+            } else {
+                self.min = std::cmp::min(self.min, key);
+                self.max = std::cmp::max(self.max, key + 1);
+            }
+            self.count += 1;
+        }
+    }
+
+    pub fn remove(&mut self, key: &u64) {
+        self.check_range(*key);
+        let address = self.get_address(key);
+        let value = self.bits.get(address);
+        if value {
+            self.count -= 1; // this will underflow if it tries to go below 0. That is like an assert.
+            self.bits.set(address, false);
+            self.purge(key);
+        }
+    }
+
+    // after removing 'key' where 'key' = min, make min the correct new min value
+    fn purge(&mut self, key: &u64) {
+        if key == &self.min && self.count > 0 {
+            let start = self.min;
+            for key in start..self.max {
+                if self.bits.get(key) {
```

Comment:
Can it use this:
https://docs.rs/bitvec/0.22.1/bitvec/slice/struct.BitSlice.html#method.leading_zeros

Reply:
These docs are from the [bitvec crate](https://crates.io/crates/bitvec). We use [this crate](https://tov.github.io/bv-rs/bv/struct.BitVec.html) elsewhere in the repo. bv does not appear to support leading_zeros.

Reply:
I see. In that case https://tov.github.io/bv-rs/bv/trait.Bits.html#method.get_block combined with ffs might be more efficient.